### PR TITLE
bpo-46921: Vectorcall support for `super()`

### PR DESCRIPTION
**Pull request by Fidget-Spinner**
Benchmarks todo.

https://bugs.python.org/issue46921

### CODE REVIEW

File: Objects/typeobject.c [hunk 1]
```diff
@@ -9072,6 +9081,45 @@ super_traverse(PyObject *self, visitproc visit, void *arg)
     return 0;
 }
 
+static PyObject *
+super_vectorcall(PyObject *self, PyObject *const *args,
+    size_t nargsf, PyObject *kwnames)
+{
+    assert(PyType_Check(self));
+    if (kwnames != NULL) {
```
Comment (corona10) [inline Objects/typeobject.c]:
```suggestion
    if (!_PyArg_NoKwnames("super", kwnames)) {
        return NULL;
    }
```

File: Objects/typeobject.c [hunk 2]
```diff
@@ -9114,4 +9161,5 @@ PyTypeObject PySuper_Type = {
     PyType_GenericAlloc,                        /* tp_alloc */
     PyType_GenericNew,                          /* tp_new */
     PyObject_GC_Del,                            /* tp_free */
+    .tp_vectorcall = (vectorcallfunc)super_vectorcall,
```
Comment (corona10) [inline Objects/typeobject.c:9167]: Would you like to apply modern C style initialization for all attributes? (tp_new, tp_alloc, ....)

File: Objects/typeobject.c [hunk 3]
```diff
@@ -9072,6 +9081,44 @@ super_traverse(PyObject *self, visitproc visit, void *arg)
...
+
+    PyTypeObject *type = NULL;
+    PyObject *obj = NULL;
+    PyTypeObject *self_type = (PyTypeObject *)self;
+    Py_ssize_t nargs = PyVectorcall_NARGS(nargsf);
+    PyObject *su = self_type->tp_alloc(self_type, 0);
+    if (su == NULL) {
+        return NULL;
+    }
+    if (nargs != 0) {
```
Comment (corona10) [inline Objects/typeobject.c:9104]:
```suggestion
     if (!_PyArg_CheckPositional("super()", nargs, 0, 2)) {
         goto fail;
    }
    if (nargs != 0) {
```

**Review by corona10 — CHANGES_REQUESTED** [review]
The current PR does not check the number of arguments.
Please add the unit test for this also :)

### AS-IS
`````
>>> super(int, int, int)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: super() takes at most 2 arguments (3 given)
`````

### PR
`````
>>> super(int, int, int)
<super: <class 'int'>, NULL>
`````

**Comment by corona10** [comment]
See my benchmark: https://bugs.python.org/msg414577
It looks like worth applying the vector call :)

**Comment by Fidget-Spinner** [comment]
@corona10 thanks for taking the time to benchmark this and for the extremely useful suggestions too. I forgot all the cool argument checking helpers we have since I'm a little rusty.

If you're interested, there's the monster GH-30992 too where I measured >2X speedup. But it's very complex and I don't have high hopes for it being merged.

File: Objects/typeobject.c (hunk 2 above)
Comment (Fidget-Spinner) [reply Objects/typeobject.c:9167]: I usually leave it as is. Some people dislike it when I convert the other attributes. I'm aware some people want to convert these too.

Comment (corona10) [reply Objects/typeobject.c:9167]: https://github.com/python/cpython/pull/30820#r790278098

At least, I and Victor love to use it :) But it's up to you

**Comment by sweeneyde** [comment]
I wonder how much a free list would help? I'd bet super objects typically have short lifetimes and not many are alive at once.

**Comment by corona10** [comment]
@Fidget-Spinner 

> Please add the unit test for this also :)

As I wrote, please add the test for checking TypeError when the given number of arguments are greater equal than 3 :)

**Comment by Fidget-Spinner** [comment]
> @Fidget-Spinner
> 
> > Please add the unit test for this also :)
> 
> As I wrote, please add the test for checking TypeError when the given number of arguments are greater equal than 3 :)

🤦 my bad, I missed that. Thanks again.

 I added one more test since I noticed it wasn't covered in the test suite.

**Comment by Fidget-Spinner** [comment]
> I wonder how much a free list would help? I'd bet super objects typically have short lifetimes and not many are alive at once.

There will probably be some improvement versus relying on CPython's obmalloc "free list". My final goal is to not need *any* super object at all though :). BTW, are you able to guesstimate how much more complexity we need for a super free list? If it isn't too complex, it's likely more worth it than my cached superinstruction overkill implementation.

**Comment by sweeneyde** [comment]
> > I wonder how much a free list would help? I'd bet super objects typically have short lifetimes and not many are alive at once.
> 
> There will probably be some improvement versus relying on CPython's obmalloc "free list". My final goal is to not need *any* super object at all though :). BTW, are you able to guesstimate how much more complexity we need for a super free list? If it isn't too complex, it's likely more worth it than my cached superinstruction overkill implementation.

I think it's (relatively) straightforward, see floatobject.c for an example. The actual free list gets attached to _PyInterpreterState there.

**Review by corona10 — APPROVED** [review]
LGTM
Feel free to merge this PR :)

**Comment by Fidget-Spinner** [comment]
@corona10 thanks for the reviews!

**Pull request merged**